### ISSUE TITLE
default component integration test blueprint contains type errors

### ISSUE THREAD

**Issue by jrr** [typed-ember/ember-cli-typescript#245]
So, I just had this experience:

```
ember g component foo-bar
installing component
  create app/components/foo-bar.ts
  create app/templates/components/foo-bar.hbs
installing component-test
  create tests/integration/components/foo-bar-test.ts
ember b
Could not start watchman
Visit https://ember-cli.com/user-guide/#watchman for more info.
tests/integration/components/foo-bar-test.ts(15,18): error TS2531: Object is possibly 'null'.
tests/integration/components/foo-bar-test.ts(24,18): error TS2531: Object is possibly 'null'.
tests/integration/components/foo-bar-test.ts(15,18): error TS2531: Object is possibly 'null'.
tests/integration/components/foo-bar-test.ts(24,18): error TS2531: Object is possibly 'null'.
cleaning up...
```

The problem is this line:

https://github.com/typed-ember/ember-cli-typescript/blob/f0ebdf5b0ab2067e2b824973e0bef808bbaf4a68/blueprints/component-test/qunit-rfc-232-files/tests/__testType__/__path__/__test__.ts#L15

They type of textContent is `string | null`. I'm not sure exactly under what circumstances it would be null, but if it is, then I would expect the exception to be caught and become a test failure.

So, given that this thing that typescript is warning about could become an exception, _and that exception is expected in the normal course of execution_, what can we do?

Is there a best practice here that has enough consensus to become the new blueprint default?

I'd prefer not to:
- lie with the typings
- null-check all over the place in my tests

Maybe we need a new set of ts-native test helpers?

**Comment by mike-north** [comment]
@jrr sometimes your two goals
> * lie with the typings
> * null-check all over the place in my tests

Will be at odds with eachother. In this situation, here's what's going on

`this.element` if of type `Element` (not necessarily `HTMLElement`), which extends `Node`, the interface with the `textContent` property (https://github.com/Microsoft/TypeScript/blob/fe387cc7f0739256337771e62398a565fbcf82c1/lib/lib.dom.d.ts#L10292)

There are all sorts of `Element`s that will never have a defined `textContent` (i.e., an `<img />` or anything else that is only usable as a self-closing tag). The boundary element of a component could concievably be one of these tags, so in a test blueprint, we cannot make assumptions one way or another. Thus, our choices are

* Ask users to test against a type that might be undefined (`"null-check all over the place in my tests"`)
* Represent remove the possibility that this value may be undefined (`"lie with the typings"`)

One way to get around this is to write tests in such a way that's tolerant of all values, without forcing tons of casting. I typically do something like

```ts
 assert.equal( ('' + this.element.textContent).trim(), ''); 
```
Alternatively, we could create some TS-friendly test helpers, which don't rely on prototype methods so much (and are thus more tolerant of `T | undefined` than `.trim()`)

```ts
function trimmed(s: string | undefined) {
  if (!s) return '';
  return s.trim();
}
...
assert.equal(trimmed(this.element.textContent), '');
```

**Comment by mike-north** [comment]
Blueprint stuff lives in [another repo](https://github.com/typed-ember/ember-cli-typescript-blueprints) now, and I believe https://github.com/typed-ember/ember-cli-typescript-blueprints/issues/4 covers the problems you've raised here.

Closing as duplicate
